### PR TITLE
Add expr ai test

### PR DESCRIPTION
**Pull request by F1r3w477**
### Problem
The Skript project has low test coverage for its existing syntax, which increases the risk of regressions. The core ExprAI syntax for getting and setting an entity's AI status had no automated tests.

### Solution
This PR introduces a test suite for ExprAI. The new test script verifies the following aspects of the expression:

Getting and setting the AI for single entities.

Handling lists of multiple entities.

Syntax aliases and patterns.

Boolean aliases.

Edge cases like setting to an unset value and querying a dead entity.

### Testing Completed
Added the ExprAI.sk syntax test. The full project test suite was run via .\gradlew clean build and passed successfully.

### Supporting Information
An interesting edge case was found during testing: querying the AI of an entity after it is killed returns true, not <none>. The added test currently asserts for this observed behavior. This is open for discussion on whether this is the desired final functionality.

---
**Completes:** none
**Related:** #6158

### CODE REVIEW

**Comment by F1r3w477** [comment]
I just want to make sure the supporting information is noticed on this PR. The added test currently asserts for the observed behavior. 

If this behavior **isn't** expected, let me know and I can open an issue for it.

**Comment by F1r3w477** [comment]
nothing like finding a bug in your own pr... sigh...

**Comment by NotSoDelayed** [comment]
> nothing like finding a bug in your own pr... sigh...

We all starts somewhere, besides I believe this is your first time contributing, so you attempting to make Skript better is a chad move 💪

File: src/test/skript/tests/syntaxes/expressions/ExprAI.sk [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+	assert ai of {_cow} is true
+
+test "Setting AI of a dead entity":
+	spawn a cow at location(40, 100, 40)
+	set {_cow} to last spawned cow
+
+	kill {_cow}
+	set ai of {_cow} to false
+
+	assert ai of {_cow} is false
```
Comment (NotSoDelayed) [inline src/test/skript/tests/syntaxes/expressions/ExprAI.sk]:
```suggestion
	assert ai of {_cow} is false
	
```
file must end with a newline (unless GitHub preview is tripping on me)

**Review by NotSoDelayed — CHANGES_REQUESTED** [review]
Suggestion on adding checks of interacting with an undefined variable
```ts
set ai of {_undefined} to false  # nothing should happen without errors
set {_value} to ai of {_undefined}  # variable stays undefined
```

File: src/test/skript/tests/syntaxes/expressions/ExprAI.sk (hunk 1 above)
Comment (F1r3w477) [reply src/test/skript/tests/syntaxes/expressions/ExprAI.sk]: fixed in latest

**Comment by F1r3w477** [comment]
oops, didn't mean to request review yet

I'll add the undefined var stuff

File: src/test/skript/tests/syntaxes/expressions/ExprAI.sk [hunk 2]
```diff
@@ -0,0 +1,63 @@
+test "Entity AI Expression Functionality":
+	spawn a cow at location(0, 100, 0)
```
Comment (Absolutionism) [inline src/test/skript/tests/syntaxes/expressions/ExprAI.sk]:
```suggestion
	spawn a cow at test-location
```
We usually limit the testing area to this

File: src/test/skript/tests/syntaxes/expressions/ExprAI.sk [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+	set {_cow} to last spawned cow
+
+	assert ai of {_cow} is true
```
Comment (Absolutionism) [inline src/test/skript/tests/syntaxes/expressions/ExprAI.sk]:
```suggestion
	assert ai of {_cow} is true with "Cow should have artificial intelligence enabled after spawn"
```
Assertions need to be accompanied with a message `with "..."`

File: src/test/skript/tests/syntaxes/expressions/ExprAI.sk [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+
+	set artificial intelligence of {_cow} to true
+	assert artificial intelligence of {_cow} is true
+
+	set ai of {_cow} to no
+	assert ai of {_cow} is false
+	set ai of {_cow} to yes
+	assert ai of {_cow} is true
+
+	kill {_cow}
```
Comment (Absolutionism) [inline src/test/skript/tests/syntaxes/expressions/ExprAI.sk]:
```suggestion
	clear entity within {_cow}
```

File: src/test/skript/tests/syntaxes/expressions/ExprAI.sk [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+
+test "Entity AI with multiple entities":
+	spawn 2 cows at location(10, 100, 10)
+	set {_cows::*} to last spawned cows
```
Comment (Absolutionism) [inline src/test/skript/tests/syntaxes/expressions/ExprAI.sk]:
```suggestion
	spawn 2 cows at test-location:
		add entity to {_cows::*}
```
You can't get multiple entities from `last spawned`

File: src/test/skript/tests/syntaxes/expressions/ExprAI.sk [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+
+	set ai of {_cows::*} to false
+	loop {_cows::*}:
+		assert ai of loop-value is false
+
+	kill {_cows::*}
```
Comment (Absolutionism) [inline src/test/skript/tests/syntaxes/expressions/ExprAI.sk]:
```suggestion
	clear entities within {_cows::*}
```

File: src/test/skript/tests/syntaxes/expressions/ExprAI.sk [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+
+test "Setting AI of a dead entity":
+	spawn a cow at location(40, 100, 40)
+	set {_cow} to last spawned cow
+
+	kill {_cow}
+	set ai of {_cow} to false
+
+	assert ai of {_cow} is false
+
```
Comment (Absolutionism) [inline src/test/skript/tests/syntaxes/expressions/ExprAI.sk]:
```suggestion

```
last line needs to have no indentation

**Review by Absolutionism — CHANGES_REQUESTED** [review]

File: src/test/skript/tests/syntaxes/expressions/ExprAI.sk (hunk 5 above)
Comment (F1r3w477) [reply src/test/skript/tests/syntaxes/expressions/ExprAI.sk]: would `add entity` add all cows? My concern being singularity and plurality of the language. If I do `add entities` would that correctly add entities/is that allowed? 

Edit: It runs fine with `entities` so I'm assuming the plurality works fine, and it's makes sense that way so I'll stick with it.

**Comment by Absolutionism** [comment]
Also another thing, it's not required, but it's usually a safer bet to do
```applescript
spawn a cow at test-location:
	set {_cow} to entity
```
rather than
```applescript
spawn a cow at test-location
set {_cow} to last spawned entity
```

File: src/test/skript/tests/syntaxes/expressions/ExprAI.sk (hunk 5 above)
Comment (Absolutionism) [reply src/test/skript/tests/syntaxes/expressions/ExprAI.sk]: The section gets ran for each entity that is spawned.

Comment (F1r3w477) [reply src/test/skript/tests/syntaxes/expressions/ExprAI.sk]: Hmm, the test fails with `entity` instead of `entities` -- Doesn't seem to be adding them to the list as `entity` maybe that functionality changed recently?

Comment (Absolutionism) [reply src/test/skript/tests/syntaxes/expressions/ExprAI.sk]: Works fine for me:
```applescript
test "Entity AI with multiple entities":
	spawn 2 cows at test-location:
		add entity to {_cows::*}

	set ai of {_cows::*} to false
	assert ai of {_cows::*} is false with "AI for all entities in a list should be set to false"

	clear entities within {_cows::*}
```

File: src/test/skript/tests/syntaxes/expressions/ExprAI.sk [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+	loop {_cows::*}:
+		assert ai of loop-value is false with "AI for all entities in a list should be set to false"
+
+	clear entities within {_cows::*}
+
+test "Setting AI to an unset value":
+	spawn a cow at test-location:
+		set {_cow} to entity
+	
+	clear {_undefined}
```
Comment (Absolutionism) [inline src/test/skript/tests/syntaxes/expressions/ExprAI.sk]:
```suggestion
```
You haven't set it, so no reason to clear it. It will still be none

File: src/test/skript/tests/syntaxes/expressions/ExprAI.sk [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+		set {_cow} to entity
+
+	clear entity within {_cow}
+	set ai of {_cow} to false
+
+	assert ai of {_cow} is false with "It should be possible to set the AI of a dead entity"
+
+test "Interacting with an undefined entity":
+	clear {_undefined}
+	clear {_value}
```
Comment (Absolutionism) [inline src/test/skript/tests/syntaxes/expressions/ExprAI.sk]:
```suggestion
```

File: src/test/skript/tests/syntaxes/expressions/ExprAI.sk [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+
+	set ai of {_undefined} to false
```
Comment (Absolutionism) [inline src/test/skript/tests/syntaxes/expressions/ExprAI.sk]:
```suggestion
```
This ultimately doesn't accomplish anything since we know `{_undefined}` is not set. No parse error or runtime  error will be thrown neither.

File: src/test/skript/tests/syntaxes/expressions/ExprAI.sk (hunk 5 above)
Comment (F1r3w477) [reply src/test/skript/tests/syntaxes/expressions/ExprAI.sk]: It parses fine for me, but when the assertion runs it receives `<none>` instead of false for the Ai value for me

**Review by Absolutionism — CHANGES_REQUESTED** [review]

File: src/test/skript/tests/syntaxes/expressions/ExprAI.sk (hunk 5 above)
Comment (Absolutionism) [reply src/test/skript/tests/syntaxes/expressions/ExprAI.sk]: What I posted passed the test on my end.

Comment (F1r3w477) [reply src/test/skript/tests/syntaxes/expressions/ExprAI.sk]: Hmm, odd. I had the same code as well. Is there a reason not to use entities?

For pure legibility sake it makes the test more readable.

Comment (Absolutionism) [reply src/test/skript/tests/syntaxes/expressions/ExprAI.sk]: Overlooked byproducts of other tests may not clear any of the entities that they spawn.
Specifically controlling the entities you spawn and manually adding to the list variable, guarantees no corruption.

Comment (F1r3w477) [reply src/test/skript/tests/syntaxes/expressions/ExprAI.sk]: If I understand correctly `entities` in this context adds all entities at the location to the list but `entity` should only apply to the entities spawned during this specific test?

File: src/test/skript/tests/syntaxes/expressions/ExprAI.sk (hunk 10 above)
Comment (F1r3w477) [reply src/test/skript/tests/syntaxes/expressions/ExprAI.sk]: I would argue that, at the very least, it verifies that no parse error or runtime error will be thrown in a case like this.

File: src/test/skript/tests/syntaxes/expressions/ExprAI.sk (hunk 5 above)
Comment (Absolutionism) [reply src/test/skript/tests/syntaxes/expressions/ExprAI.sk]: Using `entities` retrieves every entity that is spawned. Using the spawn section, uses the event-value `event-entity` but can be just used as `entity`.
For each entity that is spawned the section will get ran for each one
```applescript
spawn a cow, a pig and a chicken at test-location:
	if entity is a pig:
		set {_pig} to entity
	else if event-entity is a cow:
		set {_cow} to entity
	else if entity is a chicken:
		clear event-entity
```

File: src/test/skript/tests/syntaxes/expressions/ExprAI.sk (hunk 10 above)
Comment (Absolutionism) [reply src/test/skript/tests/syntaxes/expressions/ExprAI.sk]: If a parse error was to be thrown, what you have now doesn't properly handle it and will make the tests fail.
Plus variables don't really set off any parse errors as they can be anything during parse time.

Only thing for the near future is type hints with local variables.

Comment (F1r3w477) [reply src/test/skript/tests/syntaxes/expressions/ExprAI.sk]: Ah, I see what you are saying.

Yeah, I think this testing is out of scope for this test file then. It should be handled elsewhere, wherever we test setting vars to different things. Appreciate the conversation!

File: src/test/skript/tests/syntaxes/expressions/ExprAI.sk (hunk 5 above)
Comment (F1r3w477) [reply src/test/skript/tests/syntaxes/expressions/ExprAI.sk]: not sure what was wrong, but it's working consistently now with `entity` 🤷

**Review by Absolutionism — APPROVED** [review]

File: src/test/skript/tests/syntaxes/expressions/ExprAI.sk [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+	assert ai of {_cow} is true with "AI of a dead entity should report its last known state (true)"
+
+test "Setting AI of a dead entity":
+	spawn a cow at test-location:
+		set {_cow} to entity
+
+	clear entity within {_cow}
+	set ai of {_cow} to false
+
+	assert ai of {_cow} is false with "It should be possible to set the AI of a dead entity"
```
Comment (sovdeeth) [inline src/test/skript/tests/syntaxes/expressions/ExprAI.sk]: I would remove this test. We don't really care whether you can set the ai of a dead entity, it's basically undefined behavior. Better to not have this test since we don't guarantee that and in case it changes upstream we don't want tests to suddenly fail.

**Review by sovdeeth — CHANGES_REQUESTED** [review]

**Review by sovdeeth — APPROVED** [review]

**Review by NotSoDelayed — APPROVED** [review]

**Review by Efnilite — APPROVED** [review]
looks good
